### ISSUE TITLE
How to customize a sentenceBlot/paragraphBlot and insert  it into Quill?

### ISSUE THREAD

**Issue by johnking** [quilljs/quill#1581]
Hi there, @jhchen  
I have a use case/requirement and need some help on the implementation with Quill.

I want to create a document from a `JSON` file, the created `quill` document model need to have following  document structure in order for Application to fine control the presentation of the document with other media such as audio:
 
```
   Page -> Paragraph(s) -> sentence(s) -> words. 
```
Each element (Page, Paragraph, Sentence, Word) has its own attributes/metadata such as start-time, end-time etc.

For `Page`, I can use `scroll blot` of `quill` to represent it. 
For `Word` of each sentence, I can use `text blot` to represent it. 

But I have to customize two blots to represent sentences and paragraphs:
 
1. I want to create a sentence blot based on `inline` blot, which host all the words of it. 
2. I want to create a paragraph blot based on `block` blot, which host all the sentences of it. 

The challenge is I don't know how to insert wordBlot into sentenceBlot, insert sentenceBlot into paragraphBlot and insert paragraphBlot into quill editor. There are only insertText and insertEmbed APIs available. 

May you please give me any suggestions or hints?  

thanks a lot.
-John

**Comment by mbv-og** [comment]
Hi @johnking 

my custom **List** and **ListItem** classes replace the default formats, allowing me to use the pre-defined toolbar button to create one.

A button won't do, of course, when parsing text from a JSON object. If all you got is text, `insertEmbed` is also not what you need, because that is more for embedded content such as images (that's what you can see in the example as well: https://quilljs.com/docs/api/#insertembed).
`insertText` is actually one of the functions you can look into. https://quilljs.com/docs/api/#inserttext  The third parameter here is *format*, which you can use to .. well, format your text, e.g. in case of my Collapsible Text it could look something like this:
```es6
quill.insertText(quill.getSelection().index, 'Some nice random Text', 'collapsible', 'api');
```
At least as far as I know that should work; I don't have the means to test it right now, mind you.

You could also look into creating your own, new instance of **Delta** and use its `.insert()` functionality to the same end (documentation for that can be found here: https://github.com/quilljs/delta#insert). As far as I understand it you could even create several Deltas, one for each part (words, sentences, paragraphs and page) when iterating over the JSON Object and in the end create a single Delta out of all those using https://github.com/quilljs/delta#compose which can then be inserted into the editor using `setContents`.

I could be wrong, of course. My Collapsible Text was the first time I created a format module for Quill so I am definitely not even close to being an expert on all this. I basically took a look at the List and Code formats as an example and worked from there.

Just in case you'd like to have a look at it I created a gist with my List and ListItem here (ListItem needs to be included first, but Gist likes to sort by alphabet, apparently; in ES6 because I only have to support the latest FF and Chrome)
https://gist.github.com/mbv-og/7c717e7c9a0673a9afca24d0866185be

All of the comments in my Collapsible Text module are written in German though, so not sure how much help that would be.

Hope all this helps in some way. Sorry I cannot be of more assitance right now.

**Comment by johnking** [comment]
Hi @mbv-og , 

thanks a bunch for your reply and suggestions. I will look into your suggestions soon. I did not expect the challenge to customize a sentenceBlot and add it into quill editor. 

  I found `slate` editor very attractive but unfortunately it is integrated with `react`, and I am using `angular` instead.

  I will have a try based on your suggestion, if failed or too challenging, I will try other editor. 

thanks again.

-John
